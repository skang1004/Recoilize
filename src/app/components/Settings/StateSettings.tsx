import React from 'react';

interface StateSettingsProps {
  checked: boolean;
  setChecked: React.Dispatch<React.SetStateAction<boolean>>;
}

const StateSettings: React.FC<StateSettingsProps> = ({checked, setChecked}) => {
  // functionality to postMessage the selected snapshot index to background.js
  const persistStateFunc = (): void => {
    // setChecked as true or false
    checked ? setChecked(false) : setChecked(true);
    // variable to store/reference connection
    const backgroundConnection = chrome.runtime.connect();
    // post the message with index in payload to the connection
    backgroundConnection.postMessage({
      action: 'persistState',
      tabId: chrome.devtools.inspectedWindow.tabId,
    });
  };
  return (
    <div data-testid="stateSettings">
      <h2>Persist State</h2>
      <div className="persistContainer">
        <label className="switch" htmlFor="checkbox">
          <input
            data-testid="stateSettingsToggle"
            id="checkbox"
            type="checkbox"
            checked={checked}
            onChange={persistStateFunc}></input>
          <div className="slider round" />{' '}
        </label>
        <span className="persistText">Slide to Toggle Persist Mode</span>
      </div>
    </div>
  );
};

export default StateSettings;
