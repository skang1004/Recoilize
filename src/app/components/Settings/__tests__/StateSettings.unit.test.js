import React from 'react';
import ReactDOM from 'react-dom';
import StateSettings from '../StateSettings';

import {render, cleanup, getByTestId, fireEvent} from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

afterEach(cleanup);

const mockProps = {
  checked: false,
  setChecked: jest.fn(),
};

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<StateSettings />, div);
});

it('renders persist state text correctly', () => {
  const {getByTestId} = render(<StateSettings />);
  expect(getByTestId('stateSettings')).toHaveTextContent('Persist State');
});

describe('input checkbox', () => {
  beforeAll(cleanup);
  afterAll(cleanup);

  it('is rendered correctly', () => {
    const {getByTestId} = render(<StateSettings />);
    expect(getByTestId('stateSettingsToggle')).toHaveAttribute(
      'type',
      'checkbox',
    );
  });

  it('is unchecked initially', () => {
    const {getByTestId} = render(<StateSettings checked={mockProps.checked} />);
    const toggle = getByTestId('stateSettingsToggle');
    expect(toggle).toHaveProperty('checked', false);
  });

  it('is checked after user clicks', () => {
    const {getByTestId} = render(<StateSettings {...mockProps} />);
    const toggle = getByTestId('stateSettingsToggle');
    chrome.runtime.connect = function () {
      return {postMessage: jest.fn()};
    };
    chrome.devtools = {inspectedWindow: {}};
    fireEvent.change(toggle, {target: {checked: true}});
    expect(toggle).toBeChecked();
  });
});

it('should match snapshot when no props are passed in', () => {
  const {asFragment} = render(<StateSettings />);
  expect(asFragment()).toMatchSnapshot();
});

it('should match snapshot when props are passed in', () => {
  const {asFragment} = render(<StateSettings {...mockProps} />);
  expect(asFragment()).toMatchSnapshot();
});
