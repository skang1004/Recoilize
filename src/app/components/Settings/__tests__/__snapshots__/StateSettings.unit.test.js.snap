// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot when no props are passed in 1`] = `
<DocumentFragment>
  <div
    data-testid="stateSettings"
  >
    <h2>
      Persist State
    </h2>
    <div
      class="persistContainer"
    >
      <label
        class="switch"
        for="checkbox"
      >
        <input
          data-testid="stateSettingsToggle"
          id="checkbox"
          type="checkbox"
        />
        <div
          class="slider round"
        />
         
      </label>
      <span
        class="persistText"
      >
        Slide to Toggle Persist Mode
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`should match snapshot when props are passed in 1`] = `
<DocumentFragment>
  <div
    data-testid="stateSettings"
  >
    <h2>
      Persist State
    </h2>
    <div
      class="persistContainer"
    >
      <label
        class="switch"
        for="checkbox"
      >
        <input
          data-testid="stateSettingsToggle"
          id="checkbox"
          type="checkbox"
        />
        <div
          class="slider round"
        />
         
      </label>
      <span
        class="persistText"
      >
        Slide to Toggle Persist Mode
      </span>
    </div>
  </div>
</DocumentFragment>
`;
